src/core/ddsi/src/q_entity.c

```diff
@@ -532,6 +582,61 @@ static void add_security_builtin_endpoints(struct participant *pp, ddsi_guid_t *
 }
 #endif
 
+
+#ifdef DDSI_INCLUDE_SECURITY
+
+static void connect_participant_secure(struct q_globals *gv, struct participant *pp)
+{
+  struct proxy_participant *proxypp;
+  struct ephash_enum_proxy_participant it;
+  struct ddsi_handshake *handshake;
+  nn_wctime_t timestamp;
+
+  if (q_omg_participant_is_secure(pp))
+  {
+    timestamp = now();
+    ephash_enum_proxy_participant_init (&it, gv->guid_hash);
+    while ((proxypp = ephash_enum_proxy_participant_next (&it)) != NULL)
+    {
+      /* Do not start handshaking when security info doesn't match. */
+      if ((proxypp->handshake_admin) && (q_omg_is_similar_participant_security_info(pp, proxypp)))
+      {
+        ddsi_hsadmin_lock(proxypp->handshake_admin);
+        handshake = ddsi_hsadmin_register_locked(proxypp->handshake_admin, pp, proxypp, timestamp, handshake_end_cb);
+        assert(handshake);
+        if (!handshake) {
+          DDS_CWARNING (&gv->logconfig, "Failed to create handshake (lguid="PGUIDFMT" rguid="PGUIDFMT")\n", PGUID(pp->e.guid), PGUID(proxypp->e.guid));
+        }
+        ddsi_hsadmin_unlock(proxypp->handshake_admin);
+      }
+    }
+    ephash_enum_proxy_participant_fini (&it);
+  }
+}
+
+static void disconnect_participant_secure(struct participant *pp)
+{
+  struct proxy_participant *proxypp;
+  struct ephash_enum_proxy_participant it;
+  struct q_globals *gv;
+
+  gv = pp->e.gv;
```

Comment:
I've slowly converged on writing ``struct q_globals * const gv = pp->e.gv`` instead of this more verbose layout. (I would generally prefer a more compact layout, but that's something I'm willing to overlook 😄 )